File: backend/routes/pages.js
var express = require("express");
var rateLimit = require("express-rate-limit");
var cache = require("../cache");
var config = require("../config");
var db = require("../lib/db");
var emitters = require("../services/emitters");

var log = config.log;
var getDisplayName = emitters.getDisplayName;

var router = express.Router();

var publicLimiter = rateLimit({ windowMs: 60 * 1000, max: 60, standardHeaders: true, legacyHeaders: false });

// ── CSRF token endpoint for SPA ─────────────────────────────────────────────
router.get("/api/csrf", publicLimiter, function(req, res) {
    res.json({ csrfToken: res.locals.csrfToken || req.session.csrfToken || "" });
});

// ── User session info for SPA ───────────────────────────────────────────────
router.get("/api/me", publicLimiter, function(req, res) {
    if (!req.session || !req.session.user || !req.session.user.id) {
        return res.status(401).json({ ok: false, error: "Not authenticated" });
    }
    var userId = req.session.user.id;
    var ud = cache.usersCache[userId] || {};
    var displayName = getDisplayName(userId);
    res.json({
        ok: true,
        userId: userId,
        displayName: displayName,
        role: req.session.user.role || ud.role || "user",
        shareCode: ud.shareCode || "",
        email: ud.email || "",
        mobile: ud.mobile || ""
    });
});

// ── Live/Watch token endpoints (JSON) ───────────────────────────────────────
router.get("/api/live/:token", function(req, res) {
    var entry = cache.liveTokens.get(req.params.token);
    if (!entry) return res.json({ ok: false, expired: true });
    if (entry.expiresAt && entry.expiresAt <= Date.now()) {
        cache.liveTokens.delete(req.params.token);
        db.deleteLiveToken(req.params.token).catch(function(e) { log.error({ err: e.message }, "Failed to delete expired live token"); });
        return res.json({ ok: false, expired: true, sharedBy: getDisplayName(entry.userId) });
    }
    res.json({ ok: true, token: req.params.token, sharedBy: getDisplayName(entry.userId), expired: false });
});

router.get("/api/watch/:token", function(req, res) {
    var entry = cache.watchTokens.get(req.params.token);
    if (!entry || entry.exp < Date.now()) {
        if (entry) cache.watchTokens.delete(req.params.token);
        return res.json({ ok: false, expired: true });
    }
    res.json({ ok: true, token: req.params.token });
});

module.exports = router;

File: backend/internal/api/pages.go
package api

import (
	"context"
	"database/sql"
	"net/http"
	"time"

	"kinnect-v3/internal/auth"
	"kinnect-v3/internal/cache"
	"kinnect-v3/internal/db"
)

// PagesHandler handles CSRF, Me, LiveToken, WatchToken.
type PagesHandler struct {
	cache *cache.Cache
	db    *sql.DB
}

// Csrf handles GET /api/csrf.
func (h *PagesHandler) Csrf(w http.ResponseWriter, r *http.Request) {
	sess := auth.GetSession(r)
	token := ""
	if sess != nil {
		token = sess.CsrfToken
	}
	writeJSON(w, http.StatusOK, map[string]any{"csrfToken": token})
}

// Me handles GET /api/me.
func (h *PagesHandler) Me(w http.ResponseWriter, r *http.Request) {
	sess := auth.GetSession(r)
	if sess == nil || sess.User == nil || sess.User.ID == "" {
		writeJSON(w, http.StatusUnauthorized, map[string]any{"ok": false, "error": "Not authenticated"})
		return
	}
	userID := sess.User.ID
	role := sess.User.Role
	if role == "" {
		role = "user"
	}
	ud := h.cache.GetUser(userID)
	displayName := h.cache.GetDisplayName(userID)
	shareCode := ""
	var email, mobile string
	if ud != nil {
		shareCode = ud.ShareCode
		if ud.Email != nil {
			email = *ud.Email
		}
		if ud.Mobile != nil {
			mobile = *ud.Mobile
		}
		if role == "" {
			role = ud.Role
		}
	}
	writeJSON(w, http.StatusOK, map[string]any{
		"ok":          true,
		"userId":      userID,
		"displayName": displayName,
		"role":        role,
		"shareCode":   shareCode,
		"email":       email,
		"mobile":      mobile,
	})
}

// LiveToken handles GET /api/live/{token}.
func (h *PagesHandler) LiveToken(w http.ResponseWriter, r *http.Request) {
	token := r.PathValue("token")
	entry := h.cache.GetLiveToken(token)
	if entry == nil {
		writeJSON(w, http.StatusOK, map[string]any{"ok": false, "expired": true})
		return
	}
	now := time.Now().UnixMilli()
	if entry.ExpiresAt != nil && *entry.ExpiresAt <= now {
		h.cache.DeleteLiveToken(token)
		_ = db.DeleteLiveToken(context.Background(), h.db, token)
		sharedBy := h.cache.GetDisplayName(entry.UserID)
		writeJSON(w, http.StatusOK, map[string]any{
			"ok": false, "expired": true, "sharedBy": sharedBy,
		})
		return
	}
	sharedBy := h.cache.GetDisplayName(entry.UserID)
	writeJSON(w, http.StatusOK, map[string]any{
		"ok": true, "token": token, "sharedBy": sharedBy, "expired": false,
	})
}

// WatchToken handles GET /api/watch/{token}.
func (h *PagesHandler) WatchToken(w http.ResponseWriter, r *http.Request) {
	token := r.PathValue("token")
	entry := h.cache.GetWatchToken(token)
	if entry == nil || entry.Exp < time.Now().UnixMilli() {
		if entry != nil {
			h.cache.DeleteWatchToken(token)
		}
		writeJSON(w, http.StatusOK, map[string]any{"ok": false, "expired": true})
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"ok": true, "token": token})
}
